### PR TITLE
:sparkles: Find Nodes by hostname

### PR DESCRIPTION
**Pull request by honza**
We currently support finding Nodes:

- by ProviderID when there is a cloud provider present
- by label when we use kubeadm

But if neither cloud provider nor kubeadm is present, there is no way to find the Nodes and assign a ProviderID to them.

Here, we add a new method: matching by hostname.

This depends on #2364

### CODE REVIEW

**Comment by honza** [comment]
It looks like there are some problems when pivoting. I'm looking into it now.

**Comment by honza** [comment]
I took care of the pivoting edge case. Tests should pass now. Please take a look @lentzi90 @kashifest

File: baremetal/metal3machine_manager.go [hunk 1]
```diff
@@ -1259,73 +1286,152 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
+		return errors.Wrap(err, "unable to update the target node with providerID")
+	}
+
+	return nil
+}
+
+func (m *MachineManager) getMetal3MachineHostname() string {
+	for _, address := range m.Metal3Machine.Status.Addresses {
+		if address.Type == clusterv1.MachineHostName {
+			return address.Address
```
Comment (lentzi90) [inline baremetal/metal3machine_manager.go:1318]: Will this be enough in the long run? Do we need to handle multiple hostnames?

File: baremetal/metal3machine_manager.go [hunk 2]
```diff
@@ -1259,73 +1286,152 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
+func (m *MachineManager) SetProviderID(providerID string) {
+	m.Log.Info("ProviderID set on the Metal3Machine", "providerID", providerID)
+	m.Metal3Machine.Spec.ProviderID = &providerID
+}
+
+func (m *MachineManager) SetDefaultProviderID() {
+	namespace := m.Metal3Machine.GetNamespace()
+	m3mName := m.Metal3Machine.GetName()
+
+	// Ignoring the error because we already checked for the presence of the annotation
```
Comment (lentzi90) [inline baremetal/metal3machine_manager.go:1335]: Better to check it also here. This could be called from other places in the future that may not have checked the annotation before.

**Review by lentzi90 — COMMENTED** [review]
Thanks for taking on this! This code was definitely in need of a refactor.

File: baremetal/metal3machine_manager.go [hunk 3]
```diff
@@ -1184,6 +1194,23 @@ func (m *MachineManager) SetConditionMetal3MachineToTrue(t clusterv1.ConditionTy
...
+	if m.Metal3Cluster.Spec.CloudProviderEnabled != nil && *m.Metal3Cluster.Spec.CloudProviderEnabled {
+		return true
+	}
+
+	return false
+}
+
+func (m *MachineManager) UsingKubeadm() bool {
+	// TODO: is this enough?
+	return m.Machine.Spec.Bootstrap.ConfigRef != nil
```
[the remark this answers is not in the document]
Comment (honza) [reply baremetal/metal3machine_manager.go:1211]: The problem with that is we need a way to know if we should wait for nodes with labels to show up. If we expect nodes with labels to show up, we'll keep checking until they are up. If not, then we can proceed to find nodes by hostname/ip/etc.

File: baremetal/metal3machine_manager_test.go [hunk 4]
```diff
@@ -362,7 +361,7 @@ var _ = Describe("Metal3Machine manager", func() {
 			machineMgr.SetProviderID("correct")
 
 			Expect(*bmMachine.Spec.ProviderID).To(Equal("correct"))
-			Expect(bmMachine.Status.Ready).To(BeTrue())
+			// Expect(bmMachine.Status.Ready).To(BeTrue())
```
[the remark this answers is not in the document]
Comment (honza) [reply baremetal/metal3machine_manager_test.go]: Done

File: baremetal/metal3machine_manager.go [hunk 5]
```diff
@@ -169,6 +178,23 @@ func (m *MachineManager) IsProvisioned() bool {
 	return false
 }
 
+func (m *MachineManager) IsBaremetalHostProvisioned(ctx context.Context) bool {
+	host, _, err := m.getHost(ctx)
+	if err != nil {
+		return false
+	}
+	if host == nil {
+		errMessage := "BareMetalHost not associated, requeuing"
```
[the remark this answers is not in the document]
Comment (honza) [reply baremetal/metal3machine_manager.go]: Done

File: baremetal/metal3machine_manager.go [hunk 6]
```diff
@@ -169,6 +178,23 @@ func (m *MachineManager) IsProvisioned() bool {
...
+		m.Log.Info(errMessage)
+		return false
+	}
+	if host.Status.Provisioning.State == bmov1alpha1.StateProvisioned {
+		return true
+	}
+	m.Log.Info("Provisioning BaremetalHost, requeuing")
```
[the remark this answers is not in the document]
Comment (honza) [reply baremetal/metal3machine_manager.go]: Done

File: baremetal/metal3machine_manager.go [hunk 7]
```diff
@@ -169,6 +178,23 @@ func (m *MachineManager) IsProvisioned() bool {
...
 }
 
+func (m *MachineManager) IsBaremetalHostProvisioned(ctx context.Context) bool {
+	host, _, err := m.getHost(ctx)
+	if err != nil {
+		return false
+	}
+	if host == nil {
+		errMessage := "BareMetalHost not associated, requeuing"
+		m.Log.Info(errMessage)
```
[the remark this answers is not in the document]
Comment (honza) [reply baremetal/metal3machine_manager.go]: Done

File: baremetal/metal3machine_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (honza) [reply baremetal/metal3machine_manager.go:1208]: What about this?

We can keep trying to find nodes by label while we don't have a NodeRef on the Machine even though we already have a ProviderID set on the M3M (which we need for the external node provider use case). If we find a node by label, we update the ProviderID on the M3M. Hopefully, this will get propagated to the Machine.

File: baremetal/metal3machine_manager.go (hunk 1 above)
Comment (honza) [reply baremetal/metal3machine_manager.go]: Added support for multiple hostnames

File: baremetal/metal3machine_manager.go [hunk 8]
```diff
@@ -169,6 +177,20 @@ func (m *MachineManager) IsProvisioned() bool {
...
 
+func (m *MachineManager) IsBaremetalHostProvisioned(ctx context.Context) bool {
+	host, _, err := m.getHost(ctx)
+	if err != nil {
+		return false
+	}
+	if host == nil {
+		return false
+	}
+	if host.Status.Provisioning.State == bmov1alpha1.StateProvisioned {
```
Comment (dtantsur) [inline baremetal/metal3machine_manager.go:188]: nit: `return host.Status.Provisioning.State == bmov1alpha1.StateProvisioned`

File: baremetal/metal3machine_manager.go [hunk 9]
```diff
@@ -1259,122 +1277,207 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
-			m.Log.Info(errMessage)
-			return WithTransientError(errors.New(errMessage), requeueAfter)
-		}
+	m.SetProviderID(node.Spec.ProviderID)
+	m.SetReadyTrue()
+	return nil
+}
+
+func (m *MachineManager) NodeWithMatchingProviderIDExists(ctx context.Context, clientFactory ClientGetter) bool {
+	m.Log.Info("NodeWithMatchinProviderExists")
```
Comment (dtantsur) [inline baremetal/metal3machine_manager.go:1382]: Debug leftover? Or maybe expand into a human-friendly message.

File: baremetal/metal3machine_manager.go (hunk 2 above)
Comment (dtantsur) [reply baremetal/metal3machine_manager.go:1328]: So the comment is outdated now?

File: baremetal/metal3machine_manager.go [hunk 10]
```diff
@@ -1830,29 +2001,40 @@ func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context,
...
+
 	if matchingNodeProviderID != "" {
-		*providerIDonM3M = matchingNodeProviderID
-		return 1, nil
+		nodes := validNodes[matchingNodeProviderID]
+		return nodes[0], nil
 	}
-	return matchingNodesCount, nil
+
+	return corev1.Node{}, fmt.Errorf("unable to find node with matching ProviderID")
```
Comment (dtantsur) [inline baremetal/metal3machine_manager.go:2018]: nit: unnecessary Errorf (no formatting)

File: baremetal/metal3machine_manager.go [hunk 11]
```diff
@@ -1801,22 +1904,90 @@ func (m *MachineManager) getNodesWithLabel(ctx context.Context, nodeLabel string
...
+
+	if len(matchingNodes) > 1 {
+		errMessage := fmt.Sprintf("multiple nodes matching hostnames: %s", strings.Join(metal3MachineHostnames, ", "))
+		m.Log.Info(errMessage)
+		return WithTransientError(errors.New(errMessage), requeueAfter)
+	}
+
+	node := matchingNodes[0]
+
+	m.Log.Info("found one, setting provider id on Node")
```
Comment (dtantsur) [inline baremetal/metal3machine_manager.go:1964]: nit: maybe add a node ID to logging? Also I wonder if "found one" will be clear in the context of the entire logs

File: baremetal/metal3machine_manager.go (hunk 2 above)
Comment (honza) [reply baremetal/metal3machine_manager.go]: Yes, removed

File: baremetal/metal3machine_manager.go (hunk 8 above)
Comment (honza) [reply baremetal/metal3machine_manager.go]: Done

File: baremetal/metal3machine_manager.go [hunk 12]
```diff
@@ -1259,122 +1277,207 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
+		return fmt.Errorf("failed to json.Marshal node: %w", err)
+	}
+
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, corev1.Node{})
+	if err != nil {
+		return fmt.Errorf("failed to create patch for node %q: %w", node.GetName(), err)
+	}
+	_, err = client.Nodes().Patch(ctx, nodeVar.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil {
+		return errors.Wrap(err, "unable to update the target node with providerID")
```
[the remark this answers is not in the document]
Comment (honza) [reply baremetal/metal3machine_manager.go]: Done

File: baremetal/metal3machine_manager.go (hunk 9 above)
Comment (honza) [reply baremetal/metal3machine_manager.go]: Removed

File: baremetal/metal3machine_manager.go (hunk 10 above)
Comment (honza) [reply baremetal/metal3machine_manager.go]: Done

File: baremetal/metal3machine_manager.go [hunk 13]
```diff
@@ -169,6 +177,17 @@ func (m *MachineManager) IsProvisioned() bool {
 	return false
 }
 
+func (m *MachineManager) IsBaremetalHostProvisioned(ctx context.Context) bool {
```
Comment (lentzi90) [inline baremetal/metal3machine_manager.go:180]: nit: Please add doc string

File: baremetal/metal3machine_manager.go [hunk 14]
```diff
@@ -1259,122 +1274,204 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
+	return hostnames
+}
+
+// SetProviderID sets the metal3 provider ID on the Metal3Machine.
+func (m *MachineManager) SetProviderID(providerID string) {
+	m.Log.Info("ProviderID set on the Metal3Machine", "providerID", providerID)
+	m.Metal3Machine.Spec.ProviderID = &providerID
+}
+
+func (m *MachineManager) SetDefaultProviderID() error {
```
Comment (lentzi90) [inline baremetal/metal3machine_manager.go:1321]: nit: doc string missing

File: baremetal/metal3machine_manager.go [hunk 15]
```diff
@@ -1259,122 +1274,204 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
+	if err != nil {
+		return err
+	}
+
+	providerID := fmt.Sprintf("metal3://%s/%s/%s", namespace, bmhName, m3mName)
+	m.SetProviderID(providerID)
+	return nil
+}
+
+func (m *MachineManager) getPossibleProviderIDs(ctx context.Context) (providerIDLegacy string, providerIDNew string, err error) {
```
Comment (lentzi90) [inline baremetal/metal3machine_manager.go:1335]: This one would also benefit from a doc string

File: baremetal/metal3machine_manager.go [hunk 16]
```diff
@@ -1259,122 +1274,204 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
-
-	providerIDNew := fmt.Sprintf("metal3://%s/%s/%s", namespace, bmhName, m3mName)
+	return
+}
 
-	matchingNodesCount, err := m.getMatchingNodesWithoutLabelCount(ctx, providerIDLegacy, providerIDNew, providerIDOnM3M, clientFactory)
-	if matchingNodesCount > 1 {
-		m.Log.Info("More than one node using the same providerID")
-		return errors.Wrap(err, "More than one node using the same providerID")
+func (m *MachineManager) SetProviderIDFromCloudProviderNode(ctx context.Context, clientFactory ClientGetter) error {
```
Comment (lentzi90) [inline baremetal/metal3machine_manager.go:1358]: This one would also benefit from doc string

File: baremetal/metal3machine_manager.go [hunk 17]
```diff
@@ -1259,122 +1274,204 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
 		m.Log.Info(errMessage)
+		return false
+	}
 
-		return WithTransientError(errors.New(errMessage), requeueAfter)
+	m.Log.Info("matching node found", "node", node.GetName(), "providerID", node.Spec.ProviderID)
+	return true
+}
+
+func (m *MachineManager) SetProviderIDFromNodeLabel(ctx context.Context, clientFactory ClientGetter) (success bool, err error) {
```
Comment (lentzi90) [inline baremetal/metal3machine_manager.go:1397]: doc string here also

File: baremetal/metal3machine_manager.go [hunk 18]
```diff
@@ -1801,22 +1898,90 @@ func (m *MachineManager) getNodesWithLabel(ctx context.Context, nodeLabel string
 	return nodes, nodesCount, err
 }
 
-// getMatchingNodesWithoutLabelCount tLabel gets kubernetes nodes based on their Spec.providerID field.
-func (m *MachineManager) getMatchingNodesWithoutLabelCount(ctx context.Context, providerIDLegacy, providerIDNew string, providerIDonM3M *string, clientFactory ClientGetter) (int, error) {
+func (m *MachineManager) SetNodeProviderIDByHostname(ctx context.Context, clientFactory ClientGetter) error {
```
Comment (lentzi90) [inline baremetal/metal3machine_manager.go:1901]: doc string here also

**Review by lentzi90 — COMMENTED** [review]
The controller code looks great now!
I would just want the functions in the manager code to have some doc strings.
Are we aiming to introduce this in v1.11? I'm thinking that v1.10 is a bit too close...

**Comment by honza** [comment]
> I would just want the functions in the manager code to have some doc strings.

Done

> Are we aiming to introduce this in v1.11? I'm thinking that v1.10 is a bit too close...

We have an internal deadline in a couple of weeks so I'd like to merge this sooner rather than later. Other than that, I have no opinions.

File: baremetal/metal3machine_manager.go [hunk 19]
```diff
@@ -169,6 +177,18 @@ func (m *MachineManager) IsProvisioned() bool {
 	return false
 }
 
+// IsBaremetalHostProvisioned returns true of if the provisioning state of the underlying baremetalhost is `Provisioned`.
```
Comment (kashifest) [inline baremetal/metal3machine_manager.go:180]:
```suggestion
// IsBaremetalHostProvisioned returns true if the provisioning state of the underlying baremetalhost is `Provisioned`.
```

File: baremetal/metal3machine_manager.go [hunk 20]
```diff
@@ -169,6 +177,18 @@ func (m *MachineManager) IsProvisioned() bool {
...
+func (m *MachineManager) IsBaremetalHostProvisioned(ctx context.Context) bool {
+	host, _, err := m.getHost(ctx)
+	if err != nil {
+		return false
```
Comment (kashifest) [inline baremetal/metal3machine_manager.go:184]: shouldn't we return the error also?

File: baremetal/metal3machine_manager.go [hunk 21]
```diff
@@ -169,6 +177,18 @@ func (m *MachineManager) IsProvisioned() bool {
...
+	}
+	if host == nil {
+		return false
```
Comment (kashifest) [inline baremetal/metal3machine_manager.go:187]: Shall we put some logging here as well?

File: baremetal/metal3machine_manager.go [hunk 22]
```diff
@@ -1259,122 +1275,208 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
+// getPossibleProviderIDs returns the ProviderID for this Metal3Machine in the legacy and the new format.
+func (m *MachineManager) getPossibleProviderIDs(ctx context.Context) (providerIDLegacy string, providerIDNew string, err error) {
 	namespace := m.Metal3Machine.GetNamespace()
 	m3mName := m.Metal3Machine.GetName()
 	bmhName, err := m.getBmhNameFromM3Machine()
 	if err != nil {
 		errMessage := "unable to retrieve BMH name from Metal3Machine"
 		m.Log.Info(errMessage)
-		return WithTransientError(errors.New(errMessage), requeueAfter)
+		err = errors.Wrap(err, errMessage)
```
Comment (kashifest) [inline baremetal/metal3machine_manager.go:1345]: why are we not returning `WithTransientError` anymore here ?

File: baremetal/metal3machine_manager.go [hunk 23]
```diff
@@ -1259,122 +1275,208 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
+		return
 	}
 	bmhUID, err := m.getBmhUIDFromM3Machine(ctx)
 	if err != nil {
 		errMessage := "unable to retrieve BMH UID from Metal3Machine"
 		m.Log.Info(errMessage)
-		return WithTransientError(errors.New(errMessage), requeueAfter)
+		err = errors.Wrap(err, errMessage)
```
Comment (kashifest) [inline baremetal/metal3machine_manager.go:1352]: Same question

File: baremetal/metal3machine_manager.go [hunk 24]
```diff
@@ -1259,122 +1275,208 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
-		return errors.Wrap(err, fmt.Sprintf("Found multiple target nodes with the same label: (%s)", nodeLabel))
+		return false, errors.Wrap(err, fmt.Sprintf("Found multiple target nodes with the same label: (%s)", nodeLabel))
 	}
-	var nodeVar corev1.Node
-	for _, node := range nodes.Items {
-		oldData, err := json.Marshal(node)
+
+	providerIDLegacy, providerIDNew, err := m.getPossibleProviderIDs(ctx)
+	if err != nil {
+		errMessage := "unable to retrieve BMH name from Metal3Machine"
```
Comment (kashifest) [inline baremetal/metal3machine_manager.go:1434]: I think its better to print the err itself since `m.getPossibleProviderIDs` can return errors for other case as well i.e. when BMH UID cannot be retrieved.

**Review by kashifest — COMMENTED** [review]
The change is too big for me to grasp in one go, I will dig into it again soon. left few nits/comments in line for now.

File: baremetal/metal3machine_manager.go (hunk 22 above)
Comment (honza) [reply baremetal/metal3machine_manager.go:1345]: Good question!

My thinking was that this code is now in a private helper function and we should reserve returning `WithTransientError` to public API functions that are called directly by the controller.  I'm happy to change it if you'd prefer.

File: baremetal/metal3machine_manager.go (hunk 21 above)
Comment (honza) [reply baremetal/metal3machine_manager.go:190]: Done

File: baremetal/metal3machine_manager.go (hunk 20 above)
Comment (honza) [reply baremetal/metal3machine_manager.go:186]: It makes for a nicer API on the caller side.  If we can't find a host, I think we can be pretty confident that the host isn't provisioned?

**Comment by tuminoid** [comment]
We are about to make v1.10 beta from CAPM3. Are we expecting this to land in v1.10 or can it go to v1.11, ie. merged after branching for RC a bit later? Not exactly looking forward to merging :sparkles: this late.

File: baremetal/metal3machine_manager.go (hunk 22 above)
Comment (dtantsur) [reply baremetal/metal3machine_manager.go:1348]: I agree, it looks like it's handled on the higher level.

Comment (dtantsur) [reply baremetal/metal3machine_manager.go:1348]: nit: One day, we should get rid of errors.Wrap and other bits from the deprecated 3rd party errors package.

**Comment by lentzi90** [comment]
I would prefer to wait and merge this after branching. @honza would that be okay with the internal deadline? Do you need it as part of the release or just merged to main?

File: baremetal/metal3machine_manager.go (hunk 20 above)
Comment (dtantsur) [reply baremetal/metal3machine_manager.go:186]: As long as we're handling this error somewhere else (which we probably do), I guess it's fine?

File: controllers/metal3machine_controller.go [hunk 25]
```diff
@@ -258,30 +260,67 @@ func (r *Metal3MachineReconciler) reconcileNormal(ctx context.Context,
...
-			machineMgr.SetConditionMetal3MachineToFalse(infrav1.KubernetesNodeReadyCondition, infrav1.SettingProviderIDOnNodeFailedReason, clusterv1.ConditionSeverityError, err.Error())
 			return checkMachineError(machineMgr, err,
-				"failed to set the target node providerID", errType)
+				"Failed to set default ProviderID the Metal3Machine", errType)
 		}
-		// Make sure Spec.ProviderID is set and mark the capm3Machine ready
-		machineMgr.SetProviderID(providerID)
+		machineMgr.SetReadyTrue()
+
+		errType := capierrors.UpdateMachineError
```
Comment (dtantsur) [inline controllers/metal3machine_controller.go:302]: nit: redundant variable

File: baremetal/metal3machine_manager.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lentzi90) [reply baremetal/metal3machine_manager.go]: Thanks, looks good to me

**Comment by tuminoid** [comment]
> We have an internal deadline in a couple of weeks so I'd like to merge this sooner rather than later. Other than that, I have no opinions.

Is internal deadline meaning this should be merged by then, but you don't need it to be in a release?

**Comment by honza** [comment]
@lentzi90 @tuminoid Thanks for being accommodating! I decided to carry this patch downstream for now. We are no longer dependent on this merging in time. Let's wait until branching to be on the safe side. Thanks

**Comment by tuminoid** [comment]
> @lentzi90 @tuminoid Thanks for being accommodating! I decided to carry this patch downstream for now. We are no longer dependent on this merging in time. Let's wait until branching to be on the safe side. Thanks

I've set CAPM3 v1.11 as milestone based on this.

**Comment by suhasagasthya** [comment]
Hi @honza , Can you help me what you mean by "carry this patch downstream"? I wanted to test this to fix an issue that I am currently facing where capm3 fails to set providerID - SettingProviderIDOnNodeFailed.

**Review by lentzi90 — COMMENTED** [review]
I think we are good to move forward with this now. Thanks for the patience.
/approve

File: baremetal/metal3machine_manager.go [hunk 26]
```diff
@@ -1262,122 +1281,208 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
...
 	}
-	var nodeVar corev1.Node
-	for _, node := range nodes.Items {
-		oldData, err := json.Marshal(node)
+
+	providerIDLegacy, providerIDNew, err := m.getPossibleProviderIDs(ctx)
+	if err != nil {
+		errMessage := fmt.Sprintf("unable to retrieve BMH name from Metal3Machine: %v", err)
+		m.Log.Info(errMessage)
+		return false, WithTransientError(errors.New(errMessage), requeueAfter)
```
Comment (dtantsur) [inline baremetal/metal3machine_manager.go:1442]: nit: build the error using fmt.Errorf and %w, so that the wrapped error is stored. You can use the String() of it for logging

File: baremetal/metal3machine_manager.go [hunk 27]
```diff
@@ -1805,22 +1910,91 @@ func (m *MachineManager) getNodesWithLabel(ctx context.Context, nodeLabel string
...
-		return matchingNodesCount, errors.Wrap(err, "Error creating a remote client")
+		return errors.Wrap(err, "Error creating a remote client")
+	}
+
+	metal3MachineHostnames := m.getMetal3MachineHostnames()
+
+	if len(metal3MachineHostnames) == 0 {
+		errMessage := "No machine hostname"
+		m.Log.Info(errMessage)
+		return WithTransientError(errors.New(errMessage), requeueAfter)
```
Comment (dtantsur) [inline baremetal/metal3machine_manager.go:1925]: same about Errorf
